File: tools/JavaScript/JSDemo/algo/strategies/MacdCross.js
/**
 * algo/strategies/MacdCross.js
 *
 * Trend/momentum via MACD — a strategy port of the Pine v6 "Custom MACD
 * Indicator". Uses Algo.indicators.macd, which replicates TradingView's
 * ta.macd(src, fast, slow, signal) exactly (EMA-seeded from the first bar), so
 * the crossovers line up with the chart. Long/flat only:
 *   MACD line crosses ABOVE its signal line (hist crosses 0 up)   -> go long
 *   MACD line crosses BELOW its signal line (hist crosses 0 down)  -> go flat
 *
 * The Pine histogram coloring is purely cosmetic and has no strategy meaning, so
 * it's dropped. Decides on CLOSED bars -> backtest == live. The generous `_closes`
 * buffer lets the seeded EMAs converge to the chart's full-history values.
 */
(function (global) {
    'use strict';

    const Algo = global.Algo || (global.Algo = {});
    const Strategy = Algo.Strategy;
    const I = Algo.indicators;

    class MacdCross extends Strategy {
        static get displayName() { return 'MACD Crossover'; }

        static get params() {
            return [
                { key: 'fast', label: 'Fast', type: 'int', default: 12, min: 1, title: 'Fast EMA period' },
                { key: 'slow', label: 'Slow', type: 'int', default: 26, min: 2, title: 'Slow EMA period' },
                { key: 'signal', label: 'Signal', type: 'int', default: 9, min: 1, title: 'Signal EMA period' },
                { key: 'qty', label: 'Qty', type: 'int', default: 1, min: 1, title: 'Contracts per entry' }
            ];
        }

        static get plots() {
            return [
                { key: 'hist', label: 'Histogram', type: 'histogram', color: '#90a4ae', scale: 'osc' },
                { key: 'macd', label: 'MACD', type: 'line', color: '#2962ff', scale: 'osc' },
                { key: 'signal', label: 'Signal', type: 'line', color: '#ff6d00', scale: 'osc' }
            ];
        }

        constructor(params = {}) {
            super(params);
            this.fast = Math.max(1, parseInt(params.fast, 10) || 12);
            this.slow = Math.max(this.fast + 1, parseInt(params.slow, 10) || 26);
            this.signal = Math.max(1, parseInt(params.signal, 10) || 9);
            this.qty = Math.max(1, parseInt(params.qty, 10) || 1);

            this._closes = [];
            // Generous buffer so the seeded EMAs converge to stable values.
            this._cap = this.slow + this.signal + 250;
            this._prevHist = null;
        }

        init(broker, ctx) {
            super.init(broker, ctx);
            this.log(`MACD Crossover armed: fast=${this.fast} slow=${this.slow} signal=${this.signal} qty=${this.qty}`, 'info');
        }

        _push(close) {
            if (!Number.isFinite(close)) return;
            this._closes.push(close);
            if (this._closes.length > this._cap) this._closes.shift();
        }

        warmup(bars) {
            for (const b of bars) this._push(b?.close);
            const m = I.macd(this._closes, this.fast, this.slow, this.signal);
            this._prevHist = m ? m.hist : null;
            this.log(m != null
                ? `Warmed up from ${bars.length} bars — MACD ready (matches TradingView ta.macd)`
                : `Seeded ${this._closes.length} bars; need at least 2 closes before first MACD`, 'info');
        }

        onBar(bar) {
            this._push(bar.close);
            const m = I.macd(this._closes, this.fast, this.slow, this.signal);
            if (m == null) return; // warming up
            // Strategy View: the MACD/signal/histogram the cross is read from.
            this.plot('macd', m.macd);
            this.plot('signal', m.signal);
            this.plot('hist', m.hist);

            if (this._prevHist != null) {
                const net = this.position().net;
                const crossedUp = this._prevHist <= 0 && m.hist > 0;
                const crossedDown = this._prevHist >= 0 && m.hist < 0;

                if (crossedUp && net <= 0) {
                    if (net < 0) this.flatten();
                    this.buy(this.qty, { type: 'market' });
                    this.log(`MACD cross UP @ ${bar.close} -> BUY ${this.qty}`, 'info');
                } else if (crossedDown && net > 0) {
                    this.flatten();
                    this.log(`MACD cross DOWN @ ${bar.close} -> FLATTEN`, 'info');
                }
            }
            this._prevHist = m.hist;
        }
    }

    Algo.strategies = Algo.strategies || {};
    Algo.strategies.MacdCross = MacdCross;
})(window);

File: tools/Python/PyDemo/backtest/strategies/macd_cross.py
"""backtest/strategies/macd_cross.py

Trend/momentum via MACD, long/flat only — a port of ``MacdCross.js``. Uses
``indicators.macd`` (Pine-seeded, matching TradingView ta.macd) so crossovers
line up with the chart:
  MACD line crosses ABOVE its signal (hist crosses 0 up)   -> go long
  MACD line crosses BELOW its signal (hist crosses 0 down)  -> go flat
"""

from __future__ import annotations

import math

from .base import Strategy
from .. import indicators as I


def _int(v, default):
    try:
        return int(v)
    except (TypeError, ValueError):
        return default


class MacdCross(Strategy):
    DISPLAY_NAME = "MACD Crossover"
    PARAMS = [
        {"key": "fast", "label": "Fast", "type": "int", "default": 12, "min": 1, "title": "Fast EMA period"},
        {"key": "slow", "label": "Slow", "type": "int", "default": 26, "min": 2, "title": "Slow EMA period"},
        {"key": "signal", "label": "Signal", "type": "int", "default": 9, "min": 1, "title": "Signal EMA period"},
        {"key": "qty", "label": "Qty", "type": "int", "default": 1, "min": 1, "title": "Contracts per entry"},
    ]
    PLOTS = [
        {"key": "hist", "label": "Histogram", "type": "histogram", "color": "#90a4ae", "scale": "osc"},
        {"key": "macd", "label": "MACD", "type": "line", "color": "#2962ff", "scale": "osc"},
        {"key": "signal", "label": "Signal", "type": "line", "color": "#ff6d00", "scale": "osc"},
    ]

    def __init__(self, params=None):
        super().__init__(params)
        p = self.params
        self.fast = max(1, _int(p.get("fast"), 12))
        self.slow = max(self.fast + 1, _int(p.get("slow"), 26))
        self.signal = max(1, _int(p.get("signal"), 9))
        self.qty = max(1, _int(p.get("qty"), 1))
        self._closes = []
        # Generous buffer so the seeded EMAs converge to stable values.
        self._cap = self.slow + self.signal + 250
        self._prev_hist = None

    def init(self, broker, ctx=None):
        super().init(broker, ctx)
        self.log(f"MACD Crossover armed: fast={self.fast} slow={self.slow} "
                 f"signal={self.signal} qty={self.qty}")

    def _push(self, close):
        if close is None or not math.isfinite(close):
            return
        self._closes.append(close)
        if len(self._closes) > self._cap:
            self._closes.pop(0)

    def on_bar(self, bar):
        self._push(bar["close"])
        m = I.macd(self._closes, self.fast, self.slow, self.signal)
        if m is None:
            return  # warming up
        self.plot("macd", m["macd"])
        self.plot("signal", m["signal"])
        self.plot("hist", m["hist"])

        if self._prev_hist is not None:
            net = self.position()["net"]
            crossed_up = self._prev_hist <= 0 and m["hist"] > 0
            crossed_down = self._prev_hist >= 0 and m["hist"] < 0
            if crossed_up and net <= 0:
                if net < 0:
                    self.flatten()
                self.buy(self.qty, {"type": "market"})
                self.log(f"MACD cross UP @ {bar['close']} -> BUY {self.qty}")
            elif crossed_down and net > 0:
                self.flatten()
                self.log(f"MACD cross DOWN @ {bar['close']} -> FLATTEN")
        self._prev_hist = m["hist"]
